src/proc/validator.rs

```diff
@@ -28,8 +81,242 @@ impl Validator {
+                    }
+                }
+                if var.interpolation.is_some() {
+                    return Err(GlobalVariableError::InvalidInterpolation);
+                }
+                //TODO: prevent `Uniform` storage class with `STORE` access
+                match types[var.ty].inner {
+                    crate::TypeInner::Struct { .. }
+                    | crate::TypeInner::Image {
+                        class: crate::ImageClass::Storage(_),
+                        ..
+                    } => true,
+                    _ => false,
+                }
+            }
+            crate::StorageClass::Private | crate::StorageClass::WorkGroup => {
+                if var.binding.is_some() {
+                    return Err(GlobalVariableError::InvalidBinding);
+                }
+                if var.interpolation.is_some() {
+                    return Err(GlobalVariableError::InvalidInterpolation);
+                }
+                false
+            }
+        };
+
+        if !is_storage && !var.storage_access.is_empty() {
+            return Err(GlobalVariableError::InvalidStorageAccess);
+        }
+
+        Ok(())
+    }
+
+    fn validate_function(
+        &mut self,
+        fun: &crate::Function,
+        module: &crate::Module,
+    ) -> Result<(), FunctionError> {
+        let resolve_ctx = ResolveContext {
+            constants: &module.constants,
+            global_vars: &module.global_variables,
+            local_vars: &fun.local_variables,
+            functions: &module.functions,
+            parameter_types: &fun.parameter_types,
+        };
+        self.typifier
+            .resolve_all(&fun.expressions, &module.types, &resolve_ctx)?;
+        Ok(())
+    }
+
+    fn validate_entry_point(
+        &mut self,
+        ep: &crate::EntryPoint,
+        stage: crate::ShaderStage,
+        module: &crate::Module,
+    ) -> Result<(), EntryPointError> {
+        if ep.early_depth_test.is_some() && stage != crate::ShaderStage::Fragment {
+            return Err(EntryPointError::UnexpectedEarlyDepthTest);
+        }
+        if stage == crate::ShaderStage::Compute {
+            if ep
+                .workgroup_size
+                .iter()
+                .any(|&s| s == 0 || s > MAX_WORKGROUP_SIZE)
+            {
+                return Err(EntryPointError::InvalidWorkgroupSize);
+            }
+        } else if ep.workgroup_size != [0; 3] {
+            return Err(EntryPointError::UnexpectedWorkgroupSize);
+        }
+
+        let mut bind_group_masks = [0u64; MAX_BIND_GROUPS as usize];
+        let mut location_in_mask = 0u64;
+        let mut location_out_mask = 0u64;
+        for ((var_handle, var), &usage) in module
+            .global_variables
+            .iter()
+            .zip(&ep.function.global_usage)
+        {
+            if usage.is_empty() {
+                continue;
+            }
+
+            println!("var {:?}", var); //TEMP
```

Comment:
This can be removed? :)

Reply:
wow, this is why we need reviews :)
thank you for spotting!